core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java

```diff
@@ -1486,9 +1486,27 @@ public void delete(AlluxioURI path, DeleteOptions options) throws IOException,
         LockedInodePath inodePath =
             mInodeTree.lockInodePath(lockingScheme.getPath(), lockingScheme.getMode());
         FileSystemMasterAuditContext auditContext =
-            createAuditContext("delete", path, null, inodePath.getInodeOrNull())) {
+            createAuditContext("delete", path, null, inodePath.getInodeOrNull());
+        LockedInodePathList children = !options.isRecursive() ? null
+            : mInodeTree.lockDescendants(inodePath, InodeTree.LockMode.WRITE)) {
       try {
         mPermissionChecker.checkParentPermission(Mode.Bits.WRITE, inodePath);
+        if (children != null) {
+          List<String> failedChildren = new ArrayList<>();
+          for (LockedInodePath child : children.getInodePathList()) {
+            try {
+              mPermissionChecker.checkPermission(Mode.Bits.WRITE, child);
+            } catch (AccessControlException e) {
+              failedChildren.add(e.getMessage());
+            }
+          }
+          if (failedChildren.size() > 0) {
+            // TODO(adit): do not stop entirely if insufficient permissions to some children
+            throw new DirectoryNotEmptyException(
```

Comment:
Should this exception be permission-related instead? To reach here the user must have specified the recursive option, so there isn't any requirement that the directory be empty.

Reply:
I matched the exception with what posix does. When you delete a directory and don't have permissions to child(ren) it throws a exception saying that the directory is not empty. lmk

Reply:
I just tried with HDFS and they give a permissions error in this case. I'd prefer the HDFS approach since it's closer to the real root cause, and in general we're more similar to HDFS than POSIX

Reply:
@aaudiber does hdfs delete nothing under the directory or as much as it has permissions for?

Reply:
sounds good, thanks for verifying the hdfs behavior

Reply:
I tried and it looks like they don't delete anything